sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/http/policy/BearerTokenAuthenticationChallengePolicy.java

```diff
@@ -0,0 +1,165 @@
+            List<AuthenticationChallenge> challenges = parseChallenges(authHeader);
+            for (AuthenticationChallenge authenticationChallenge : challenges) {
+                Map<String, String> extractedChallengeParams =
+                    parseChallengeParams(authenticationChallenge.getChallengeParameters());
+                if (extractedChallengeParams.containsKey(CLAIMS_PARAMETER)) {
+                    String claims = new String(Base64.getUrlDecoder()
+                        .decode(extractedChallengeParams.get(CLAIMS_PARAMETER)), StandardCharsets.UTF_8);
+                    return authenticateRequest(context,
+                        () -> credential.getToken(new TokenRequestContext()
+                                                      .addScopes(scopes).setClaims(claims)), true)
+                               .flatMap(b -> Mono.just(true));
+                }
+            }
+        }
+        return Mono.just(false);
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            return Mono.error(new RuntimeException("token credentials require a URL using the HTTPS protocol scheme"));
+        }
+        HttpPipelineNextPolicy nextPolicy = next.clone();
+
+        return onBeforeRequest(context)
+               .then(next.process())
+               .flatMap(httpResponse -> {
+                   String authHeader = httpResponse.getHeaderValue(WWW_AUTHENTICATE);
+                   if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+                       return onChallenge(context, httpResponse).flatMap(retry -> {
+                           if (retry) {
+                               return nextPolicy.process();
+                           } else {
+                               return Mono.just(httpResponse);
+                           }
+                       });
+                   }
+                   return Mono.just(httpResponse);
+               });
+    }
+
+    /**
+     * Get the {@link AccessTokenCache} holding the cached access tokens and the logic to retrieve and refresh
+     * access tokens.
+     *
+     * @return the {@link AccessTokenCache}
+     */
+    public AccessTokenCache getTokenCache() {
+        return cache;
+    }
+
+    private Mono<Void> authenticateRequest(HttpPipelineCallContext context, Supplier<Mono<AccessToken>> tokenSupplier,
+                                           boolean forceTokenRefresh) {
+        return cache.getToken(tokenSupplier, forceTokenRefresh)
+           .flatMap(token -> {
+               context.getHttpRequest().getHeaders().set(AUTHORIZATION_HEADER, BEARER + " " + token.getToken());
+               return Mono.empty();
+           });
+    }
+
+    List<AuthenticationChallenge> parseChallenges(String header) {
+        Matcher matcher = AUTHENTICATION_CHALLENGE_PATTERN.matcher(header);
```

Comment:
Instead of needing to use a regex could we use `HttpHeader.getValue` or `HttpHeader.getValueList`? It's looking like we use the regex to handle comma concatenated header values.

Reply:
this could be utilized in the next preview release, will try this out.